aws-sigv4/src/lib.rs

```diff
@@ -26,48 +28,110 @@ pub fn sign<B>(
 where
     B: AsRef<[u8]>,
 {
+    for (header_name, header_value) in sign_core(
+        &req,
+        Config {
+            access_key: &credential.access_key,
+            secret_key: &credential.secret_key,
+            security_token: credential.security_token.as_deref(),
+            region,
+            svc,
+            date: SystemTime::now(),
+        },
+    ) {
+        req.headers_mut()
+            .append(header_name.header_name(), header_value.parse()?);
+    }
+
+    Ok(())
+}
+
+/// SignatureKey is the key portion of the key-value pair of a generated SigV4 signature.
+///
+/// When signing with SigV4, the algorithm produces multiple components of a signature that MUST
+/// be applied to a request.
+pub enum SignatureKey {
+    Authorization,
+    AmzDate,
+    AmzSecurityToken,
+}
+
+impl SignatureKey {
+    pub fn header_name(&self) -> HeaderName {
+        match self {
+            SignatureKey::Authorization => header::AUTHORIZATION,
+            SignatureKey::AmzDate => HeaderName::from_static(X_AMZ_DATE),
+            SignatureKey::AmzSecurityToken => HeaderName::from_static(X_AMZ_SECURITY_TOKEN),
+        }
+    }
+}
+
+pub struct Config<'a> {
+    pub access_key: &'a str,
+    pub secret_key: &'a str,
+    pub security_token: Option<&'a str>,
+
+    pub region: &'a str,
+    pub svc: &'a str,
+
+    pub date: SystemTime,
+}
+
+pub fn sign_core<'a, B>(
+    req: &'a http::Request<B>,
+    config: Config<'a>,
+) -> impl Iterator<Item = (SignatureKey, String)>
+where
+    B: AsRef<[u8]>,
```

Comment:
I think I mentioned my worries about this, but I think it's fine and can be iterated upon.